### ISSUE TITLE
Perf testing: Cover Site Editor loading time

### ISSUE THREAD

**Issue by ockham** [WordPress/gutenberg#23793]
**Is your feature request related to a problem? Please describe.**
We've recently seen an issue (#23662) where the Site Editor took >10 seconds to load on a HyperDB-powered WordPress install that was using a distributed database. The reason was that in the PHP generating the relevant page, we were accidentally creating too many auto-drafts, which triggered HyperDB's SRTM (Send Reads To Master).

**Describe the solution you'd like**
While those are special circumstances that would easily go unnoticed on a single-database install, we realized that we could avoid similar issues by adding some basic perf test coverage to the Site Editor.

**Describe alternatives you've considered**
N/A

**Comment by ockham** [comment]
Relevant docs/files:

- https://github.com/WordPress/gutenberg/blob/710d52158514dc20a053e36ac3031305b52f6152/docs/contributors/testing-overview.md#performance-testing
- https://github.com/WordPress/gutenberg/blob/a7e56be92f4330494fbd3283b67542457094511b/packages/e2e-tests/specs/performance/performance.test.js. This is the perf test file for the Post Editor. The Jest setup file [makes it seem](https://github.com/WordPress/gutenberg/blob/a7e56be92f4330494fbd3283b67542457094511b/packages/e2e-tests/jest.performance.config.js#L3) like we could simply add another file (with a `.test.js` suffix) to that folder, but there's some [more specific tooling](https://github.com/WordPress/gutenberg/blob/a7e56be92f4330494fbd3283b67542457094511b/bin/plugin/commands/performance.js) that relies on the presence of a single `results.json` file.

**Comment by ockham** [comment]
I think that it might make sense to try and add a separate file to cover the Site Editor (e.g. `packages/e2e-tests/specs/performance/site-editor.test.js`), but we'll need to modify the other tooling accordingly.

For starters, I'd like us to cover loading time as a minimum, specifically to make sure template resolution doesn't take a long time. The actual test will then be relatively simple, but we need to consider what fixtures/assets will be required.

I think we'll need at least a very basic block-based theme in `assets` in order to properly simulate template resolution. An obvious contender would be the [Twenty Twenty Blocks](https://github.com/WordPress/theme-experiments/tree/8fe55c614f4048871534f02da01762359c224436/twentytwenty-blocks) theme from the `theme-experiments` repo -- I guess we can just copy a current snapshot to [`packages/e2e-tests/assets`](https://github.com/WordPress/gutenberg/tree/a7e56be92f4330494fbd3283b67542457094511b/packages/e2e-tests/assets).

**Comment by ockham** [comment]
We have some prior art for Site Editor e2e tests, e.g. https://github.com/WordPress/gutenberg/blob/a7e56be92f4330494fbd3283b67542457094511b/packages/e2e-tests/specs/experiments/template-part.test.js

(For measuring editor loading time, we won't need too much of the tooling used there though.)

**Comment by ockham** [comment]
We'll likely need to modify [`performance-reporter.js`](https://github.com/WordPress/gutenberg/blob/a7e56be92f4330494fbd3283b67542457094511b/packages/e2e-tests/config/performance-reporter.js), which is currently evaluating a [hard-wired](https://github.com/WordPress/gutenberg/blob/a7e56be92f4330494fbd3283b67542457094511b/packages/e2e-tests/config/performance-reporter.js#L21) `results.json`. `performance-reporter.js` is configured as a [`reporter`](https://github.com/WordPress/gutenberg/blob/a7e56be92f4330494fbd3283b67542457094511b/packages/e2e-tests/jest.performance.config.js#L15) in `jest.performance.config.js`.

Reading up on Jest's [`reporters`](https://jestjs.io/docs/en/configuration#reporters-arraymodulename--modulename-options).

**Comment by ockham** [comment]
cc/ @youknowriad since I've noticed you're working on some perf test improvements (#23817, #23818)

tl;dr We'd like to run some perf tests on the Site Editor. The above describes the state of my research and rough plan of action.

**Comment by youknowriad** [comment]
Sounds good to me 👍 

> I think we'll need at least a very basic block-based theme in assets in order to properly simulate template resolution. 

Isn't this what the demo templates are for?

**Comment by ockham** [comment]
> Sounds good to me 
> 
> > I think we'll need at least a very basic block-based theme in assets in order to properly simulate template resolution.
> 
> Isn't this what the demo templates are for?

Good point, thanks!
